### PR TITLE
HPCC-30057 Use DataBuffers to store returned rows in localAgent mode

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
Here are some before and after stats.  Median times for 9 runs of stresstext.ecl

Num          Before       After
Parallel
1, 179, 151
2, 314, 251
3, 438, 347
4, 566, 492
5, 703, 631
6, 831, 686
8, 974, 986
10, 1173, 1200
12, 1352, 1430
15, 1680, 1709
20, 2686, 2733
25, 3093, 3146
30, 3762, 3754
35, 5515, 5575
40, 7755, 7828
45, 8729, 8746
50, 8779, 8686
60, 11881, 11802

So a notable improvement for n<8.  After that something else probably becomes the contended bottleneck

**Review by mckellyln — APPROVED** [review]
I think I need more time to really go over this in detail, but overall it looks good to me.

File: roxie/ccd/ccdqueue.cpp [hunk 1]
```diff
@@ -3010,32 +3010,194 @@ unsigned CDummyMessagePacker::size() const
...
+            }
+            if (variable)
+                return ((char *) tempBuffer) + sizeof(RecordLengthType);
+            else
+                return tempBuffer;
+        }
+        else
+        {
+            // Will fit in one, though not necessarily the current one...
+            packed = false;
```
Comment (ghalliday) [inline roxie/ccd/ccdqueue.cpp:3131]: Shouldn't this be packed=true?

File: devdoc/roxie.md [hunk 2]
```diff
@@ -281,3 +281,19 @@ Should the scope of the blacklist be different? Possible scopes are:
...
+
+What is the lifespan issue? In-flight queries may be abandoned when a server-side query fails, times out, or no longer needs the data. Using DataBuffer does not have this issue as they are attached to the query's memory manager/allocation once read. Or we could bypass the agent queue altogether, but rather more refactoring needed for that (might almost be easier to extent the "local optimization" mode to use multiple threads at that point)
+
+abortPending, replyPending, and abortPendingData methods are unimplemented, which may lead to some inefficiencies?
+
+LocalMessagePacker currently derives from DummyMessagePacker, but they may be used for different things... DummyMessagePacker is used in parallel aggregate activities to intercept the output from each parallel processor so that they can be combined and resent. These don't really have the lifespan issues, and might be better off using unserialized rows in the parent activity's RowManager. Perhaps the first step is to make LocalMessagePaker independent of DummyMessagePacker by copying the functions into the former.
```
Comment (ghalliday) [inline devdoc/roxie.md:299]: update this section in the light of your changes?

**Review by ghalliday — COMMENTED** [review]
@richardkchapman first scan looks good.  One issue which I think would still work because data would be copied over itself.
Please squash (but don't rebase) with that change.

File: roxie/ccd/ccdqueue.cpp (hunk 1 above)
Comment (richardkchapman) [reply roxie/ccd/ccdqueue.cpp:3131]: Yes I think so

File: devdoc/roxie.md (hunk 2 above)
Comment (richardkchapman) [reply devdoc/roxie.md:299]: Will do

**Review by ghalliday — APPROVED** [review]

**Pull request merged**